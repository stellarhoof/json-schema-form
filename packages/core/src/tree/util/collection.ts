import isPlainObject from "lodash/isPlainObject.js"

export type Key = string | number | symbol

export type Collection<T> = Array<T> | Record<Key, T>

export const forEach = <T>(
  callback: (child: T, key: Key) => void,
  children: Collection<T>
): void => {
  if (children instanceof Array) {
    children.forEach((v, k) => callback(v, k))
  }
  if (isPlainObject(children)) {
    for (const key in children) {
      callback(children[key], key)
    }
  }
}

export const map = <T, T2 = any>(
  callback: (child: T, key: Key) => T2,
  children: Collection<T> | any
): Collection<T2> | [] => {
  if (children instanceof Array) {
    return children.map((v, k) => callback(v, k))
  }
  if (isPlainObject(children)) {
    const result: Collection<T2> = {}
    for (const key in children) {
      result[key] = callback(children[key], key)
    }
    return result
  }
  return []
}

export const filter = <T>(
  callback: (child: T, key: Key) => boolean,
  children: Collection<T> | any
): Collection<T> | [] => {
  if (children instanceof Array) {
    return children.filter((v, k) => callback(v, k))
  }
  if (isPlainObject(children)) {
    const result: Collection<T> = {}
    for (const key in children) {
      if (callback(children[key], key)) {
        result[key] = children[key]
      }
    }
    return result
  }
  return []
}
