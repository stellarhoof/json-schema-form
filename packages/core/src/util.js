import _ from "lodash/fp.js"

export const joinPaths = (...segments) =>
  _.filter((x) => x !== "" && !_.isNil(x), segments).join(".")
