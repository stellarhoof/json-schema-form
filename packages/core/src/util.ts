import _ from "lodash/fp.js"

export const joinPaths = (
  ...segments: (string | number | null | undefined)[]
) => _.filter((x) => x !== "" && !_.isNil(x), segments).join(".")
