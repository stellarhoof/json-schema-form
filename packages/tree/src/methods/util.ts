import {
  Key,
  Context,
  TraversalType,
  NodeType,
  TraversalOptions,
} from "../types.ts"

export const initialContext = { depth: 0, parents: [], parentsKeys: [] }

export const nextContext = <Node>(
  node: Node,
  key: Key,
  ctx: Context<Node>
): Context<Node> => ({
  depth: ctx.depth + 1,
  parents: [node, ...ctx.parents],
  parentsKeys: [key, ...ctx.parentsKeys],
})

export const resolveOptions = <Children>(
  children: Children,
  {
    visit = NodeType.Any,
    traversal = TraversalType.PreOrder,
  }: Partial<TraversalOptions> = {}
) => {
  const canVisit =
    visit === NodeType.Any ||
    (visit === NodeType.Leaf && !children) ||
    (visit === NodeType.Tree && children)
  return {
    canVisitPre: canVisit && traversal === TraversalType.PreOrder,
    canVisitPost: canVisit && traversal === TraversalType.PostOrder,
  }
}
