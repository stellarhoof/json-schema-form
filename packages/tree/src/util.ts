import reverse from "lodash/reverse.js"
import dropRight from "lodash/dropRight.js"
import isPlainObject from "lodash/isPlainObject.js"
import { Key, Context, Config } from "./types.ts"

export const buildPath = <Node>(_node: Node, key: Key, ctx: Context<Node>) =>
  reverse([key, ...dropRight(ctx.parentsKeys)])

export const foo = <Node extends Record<Key, any>, Children>(
  key: Key
): Pick<Config<Node, Children>, "getChildren" | "setChildren"> => ({
  getChildren(node) {
    if (isPlainObject(node) && key in node) {
      return (node as Node)[key] as Children
    }
  },
  setChildren(children, node) {
    return {
      ...node,
      // eslint-disable-next-line @typescript-eslint/no-unsafe-assignment
      [key]: children,
    }
  },
})
