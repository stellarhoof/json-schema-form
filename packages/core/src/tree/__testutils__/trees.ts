import isPlainObject from "lodash/isPlainObject.js"
import { Key, Collection } from "../util/collection.js"
import { Tree } from "../index.js"

export namespace SimpleTree {
  export const name = "SimpleTree"

  type T = any

  export const tree = new Tree<T>()

  export const getValue = (): T => ({
    title: "The Empire Strikes Back",
    planets: ["Aldeeran", { name: "Tatooine" }],
  })
}

export namespace TreeWithChildrenKey {
  export const name = "TreeWithChildrenKey"

  interface T extends Record<Key, any> {
    [index: Key]: undefined | string | Collection<T>
    children?: Collection<T>
  }

  export const tree = new Tree<T>({
    getChildren(node) {
      if (isPlainObject(node) && "children" in node) {
        return node.children
      }
    },
    setChildren(children, node) {
      return { ...node, children }
    },
  })

  export const getValue = (): T => ({
    name: "Anakin Skywalker",
    children: [{ name: "Luke Skywalker", films: [{ title: "A New Hope" }] }],
  })
}
