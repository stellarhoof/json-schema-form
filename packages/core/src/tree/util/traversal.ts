import isPlainObject from "lodash/isPlainObject.js"
import { Key, Collection } from "./collection.js"

export type Iteratee<F extends (...args: any[]) => any> =
  | symbol
  | number
  | string
  | object
  | F

export enum TraversalType {
  PreOrder = "PreOrder",
  PostOrder = "PostOrder",
}

export enum NodeType {
  Any = "Any",
  Leaf = "Leaf",
  Tree = "Tree",
}

export type Options<T> = {
  visit: NodeType
  traversal: TraversalType
  getChildren: (node: T) => Collection<T> | undefined
  setChildren: (children: Collection<T>, node: T) => T
}

export type Context<T> = {
  depth: number
  parents: T[]
  parentsKeys: Key[]
}

export const initialContext = { depth: 0, parents: [], parentsKeys: [] }

export const nextContext = <T>(
  node: T,
  key: Key,
  context: Context<T>
): Context<T> => ({
  depth: context.depth + 1,
  parents: [node, ...context.parents],
  parentsKeys: [key, ...context.parentsKeys],
})

export const resolveOptions = <T>(
  node: T,
  { visit, traversal, getChildren, setChildren }: Partial<Options<T>> = {}
) => {
  visit ??= NodeType.Any
  traversal ??= TraversalType.PreOrder
  setChildren ??= (children) => {
    return children as unknown as T
  }
  getChildren ??= (node) => {
    if (node instanceof Array || isPlainObject(node)) {
      return node as unknown as Collection<T>
    }
  }
  const children = getChildren(node)
  const canVisit =
    visit === NodeType.Any ||
    (visit === NodeType.Leaf && !children) ||
    (visit === NodeType.Tree && children)
  return {
    children,
    setChildren,
    getChildren,
    canVisitPre: canVisit && traversal === TraversalType.PreOrder,
    canVisitPost: canVisit && traversal === TraversalType.PostOrder,
  }
}
