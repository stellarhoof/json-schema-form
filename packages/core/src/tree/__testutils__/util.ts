import { SimpleTree, TreeWithChildrenKey } from "./trees.js"
import { NodeType, TraversalType } from "../util/traversal.js"

export const testCases = [
  {
    ...SimpleTree,
    options: { visit: NodeType.Leaf, traversal: TraversalType.PreOrder },
  },
  {
    ...SimpleTree,
    options: { visit: NodeType.Leaf, traversal: TraversalType.PostOrder },
  },
  {
    ...SimpleTree,
    options: { visit: NodeType.Tree, traversal: TraversalType.PreOrder },
  },
  {
    ...SimpleTree,
    options: { visit: NodeType.Tree, traversal: TraversalType.PostOrder },
  },
  {
    ...SimpleTree,
    options: { visit: NodeType.Any, traversal: TraversalType.PreOrder },
  },
  {
    ...SimpleTree,
    options: { visit: NodeType.Any, traversal: TraversalType.PostOrder },
  },
  {
    ...TreeWithChildrenKey,
    options: { visit: NodeType.Leaf, traversal: TraversalType.PreOrder },
  },
  {
    ...TreeWithChildrenKey,
    options: { visit: NodeType.Leaf, traversal: TraversalType.PostOrder },
  },
  {
    ...TreeWithChildrenKey,
    options: { visit: NodeType.Tree, traversal: TraversalType.PreOrder },
  },
  {
    ...TreeWithChildrenKey,
    options: { visit: NodeType.Tree, traversal: TraversalType.PostOrder },
  },
  {
    ...TreeWithChildrenKey,
    options: { visit: NodeType.Any, traversal: TraversalType.PreOrder },
  },
  {
    ...TreeWithChildrenKey,
    options: { visit: NodeType.Any, traversal: TraversalType.PostOrder },
  },
]
