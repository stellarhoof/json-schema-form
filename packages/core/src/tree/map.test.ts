import { it, expect } from "vitest"
import isPlainObject from "lodash/isPlainObject.js"
import { Key } from "./util/collection.js"
import { testCases } from "./__testutils__/util.js"
import { TreeWithChildrenKey } from "./__testutils__/trees.js"

const callback = (node) => {
  if (typeof node === "string") return `${node} added[string]`
  if (node instanceof Array) return [...node, "added[array]"]
  if (isPlainObject(node)) return { ...node, key: "added[object]" }
  return node
}

it.each(testCases)(
  "tree:$name visit:$options.visit traversal:$options.traversal",
  ({ tree, getValue, options }) => {
    const result = tree.map(callback, getValue(), options)
    expect(result).toMatchSnapshot()
  }
)

it("should not keep traversing when children are dynamically removed", () => {
  const { tree, getValue } = TreeWithChildrenKey
  const keys: Key[] = []

  tree.map((node, key) => {
    keys.push(key)
    if (key === "") {
      return { ...node, children: {} }
    }
  }, getValue())

  expect(keys).toEqual([""])
})

it("should keep traversing when children are dynamically added", () => {
  const { tree, getValue } = TreeWithChildrenKey
  const keys: Key[] = []

  tree.map((node, key) => {
    keys.push(key)
    if (key === "") {
      return {
        ...node,
        children: {
          related: { name: "Game of Thrones" },
        },
      }
    }
  }, getValue())

  expect(keys).toEqual(["", "related"])
})
