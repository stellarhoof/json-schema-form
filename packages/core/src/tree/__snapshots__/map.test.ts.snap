// Vitest Snapshot v1

exports[`tree:'SimpleTree' visit:'Any' traversal:'PostOrder' 1`] = `
{
  "key": "added[object]",
  "planets": [
    "Aldeeran added[string]",
    {
      "key": "added[object]",
      "name": "Tatooine added[string]",
    },
    "added[array]",
  ],
  "title": "The Empire Strikes Back added[string]",
}
`;

exports[`tree:'SimpleTree' visit:'Any' traversal:'PreOrder' 1`] = `
{
  "key": "added[object] added[string]",
  "planets": [
    "Aldeeran added[string]",
    {
      "key": "added[object] added[string]",
      "name": "Tatooine added[string]",
    },
    "added[array] added[string]",
  ],
  "title": "The Empire Strikes Back added[string]",
}
`;

exports[`tree:'SimpleTree' visit:'Leaf' traversal:'PostOrder' 1`] = `
{
  "planets": [
    "Aldeeran added[string]",
    {
      "name": "Tatooine added[string]",
    },
  ],
  "title": "The Empire Strikes Back added[string]",
}
`;

exports[`tree:'SimpleTree' visit:'Leaf' traversal:'PreOrder' 1`] = `
{
  "planets": [
    "Aldeeran added[string]",
    {
      "name": "Tatooine added[string]",
    },
  ],
  "title": "The Empire Strikes Back added[string]",
}
`;

exports[`tree:'SimpleTree' visit:'Tree' traversal:'PostOrder' 1`] = `
{
  "key": "added[object]",
  "planets": [
    "Aldeeran",
    {
      "key": "added[object]",
      "name": "Tatooine",
    },
    "added[array]",
  ],
  "title": "The Empire Strikes Back",
}
`;

exports[`tree:'SimpleTree' visit:'Tree' traversal:'PreOrder' 1`] = `
{
  "key": "added[object]",
  "planets": [
    "Aldeeran",
    {
      "key": "added[object]",
      "name": "Tatooine",
    },
    "added[array]",
  ],
  "title": "The Empire Strikes Back",
}
`;

exports[`tree:'TreeWithChildrenKey' visit:'Any' traversal:'PostOrder' 1`] = `
{
  "children": [
    {
      "films": [
        {
          "title": "A New Hope",
        },
      ],
      "key": "added[object]",
      "name": "Luke Skywalker",
    },
  ],
  "key": "added[object]",
  "name": "Anakin Skywalker",
}
`;

exports[`tree:'TreeWithChildrenKey' visit:'Any' traversal:'PreOrder' 1`] = `
{
  "children": [
    {
      "films": [
        {
          "title": "A New Hope",
        },
      ],
      "key": "added[object]",
      "name": "Luke Skywalker",
    },
  ],
  "key": "added[object]",
  "name": "Anakin Skywalker",
}
`;

exports[`tree:'TreeWithChildrenKey' visit:'Leaf' traversal:'PostOrder' 1`] = `
{
  "children": [
    {
      "films": [
        {
          "title": "A New Hope",
        },
      ],
      "key": "added[object]",
      "name": "Luke Skywalker",
    },
  ],
  "name": "Anakin Skywalker",
}
`;

exports[`tree:'TreeWithChildrenKey' visit:'Leaf' traversal:'PreOrder' 1`] = `
{
  "children": [
    {
      "films": [
        {
          "title": "A New Hope",
        },
      ],
      "key": "added[object]",
      "name": "Luke Skywalker",
    },
  ],
  "name": "Anakin Skywalker",
}
`;

exports[`tree:'TreeWithChildrenKey' visit:'Tree' traversal:'PostOrder' 1`] = `
{
  "children": [
    {
      "films": [
        {
          "title": "A New Hope",
        },
      ],
      "name": "Luke Skywalker",
    },
  ],
  "key": "added[object]",
  "name": "Anakin Skywalker",
}
`;

exports[`tree:'TreeWithChildrenKey' visit:'Tree' traversal:'PreOrder' 1`] = `
{
  "children": [
    {
      "films": [
        {
          "title": "A New Hope",
        },
      ],
      "name": "Luke Skywalker",
    },
  ],
  "key": "added[object]",
  "name": "Anakin Skywalker",
}
`;
